### PR TITLE
[FLINK-31055][runtime] Fix the bug that dynamic flag of stream graph does not take effect when translating the transformations

### PR DESCRIPTION
**Pull request by wanglijie95**
## What is the purpose of the change

Currently, the dynamic flag of stream graph is not set when [translate transformations](https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java#L324). However, the dynamic flag will be used ([here](https://github.com/apache/flink/blob/master/flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraph.java#L696)) when translating, we should set the dynamic flag before the translating.

## Verifying this change
`ForwardForUnspecifiedPartitionerTest#testConvertToCorrectPartitioner`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**no**)
  - The serializers: (**no**)
  - The runtime per-record code paths (performance sensitive): (**no**)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**no**)
  - The S3 file system connector: (**no**)

## Documentation

  - Does this pull request introduce a new feature? (**no**)
  - If yes, how is the feature documented? (**not applicable**)

### CODE REVIEW

**Review by JunRuiLee — APPROVED** [review]
LGTM！

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java [hunk 1]
```diff
@@ -316,6 +316,14 @@ public StreamGraph generate() {
...
                         ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH));
         shouldExecuteInBatchMode = shouldExecuteInBatchMode();
+        Optional<JobManagerOptions.SchedulerType> schedulerTypeOptional =
+                executionConfig.getSchedulerType();
+        boolean dynamic =
+                shouldExecuteInBatchMode
+                        && schedulerTypeOptional.orElse(
+                                        JobManagerOptions.SchedulerType.AdaptiveBatch)
+                                == JobManagerOptions.SchedulerType.AdaptiveBatch;
+        streamGraph.setDynamic(dynamic);
```
Comment (zhuzhurk) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java:326]: How about to introduce a method `maybeSetDynamic()` and invoke it in `configureStreamGraph(...)`?

Comment (zhuzhurk) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java:326]: Maybe add a hotfix commit to also move the above invocations `setAutoParallelismEnabled` and `setEnableCheckpointsAfterTasksFinish` into `configureStreamGraph()`.

Comment (wanglijie95) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java:326]: -> How about to introduce a method maybeSetDynamic() and invoke it in configureStreamGraph(...)?
Maybe `setDynamic(final StreamGraph graph)` ?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java [hunk 2]
```diff
@@ -359,6 +345,17 @@ private boolean shouldDisableUnalignedCheckpointing(StreamEdge edge) {
         return partitioner.isPointwise() || partitioner.isBroadcast();
     }
 
+    private void setDynamic(final StreamGraph graph) {
```
Comment (zhuzhurk) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java:348]: This parameter is not needed.

**Review by zhuzhurk — APPROVED** [review]
LGTM.

**Pull request closed without merging**